File: sdk/js/node/src/groups.ts
import type { HttpClient } from "./client.js";
import type { ListParams, Page } from "./users.js";

export interface Group {
  id: string;
  tenant_id: string;
  name: string;
  description?: string;
  metadata?: Record<string, unknown>;
  created_at: string;
  updated_at?: string;
}

export interface CreateGroupInput {
  name: string;
  tenant_id?: string;
  description?: string;
  metadata?: Record<string, unknown>;
}

export interface UpdateGroupInput {
  name?: string;
  description?: string;
  metadata?: Record<string, unknown>;
}

export interface GroupMember {
  user_id: string;
  group_id: string;
  added_at: string;
}

interface ListEnvelope<T> {
  items?: T[];
  data?: T[];
  next_cursor?: string;
}

export class Groups {
  constructor(private readonly http: HttpClient) {}

  create(input: CreateGroupInput): Promise<Group> {
    return this.http.post<Group>("/v1/groups", input);
  }

  get(id: string): Promise<Group> {
    return this.http.get<Group>(`/v1/groups/${encodeURIComponent(id)}`);
  }

  update(id: string, input: UpdateGroupInput): Promise<Group> {
    return this.http.patch<Group>(`/v1/groups/${encodeURIComponent(id)}`, input);
  }

  delete(id: string): Promise<void> {
    return this.http.delete<void>(`/v1/groups/${encodeURIComponent(id)}`);
  }

  async list(params: ListParams = {}): Promise<Page<Group>> {
    const res = await this.http.get<ListEnvelope<Group>>("/v1/groups", {
      query: { tenant: params.tenant, limit: params.limit, cursor: params.cursor },
    });
    return { data: res.items ?? res.data ?? [], nextCursor: res.next_cursor };
  }

  async *listAll(params: ListParams = {}): AsyncGenerator<Group> {
    let cursor = params.cursor;
    do {
      const page = await this.list({ ...params, cursor });
      for (const item of page.data) yield item;
      cursor = page.nextCursor;
    } while (cursor);
  }

  addMember(groupId: string, userId: string): Promise<void> {
    return this.http.post<void>(`/v1/groups/${encodeURIComponent(groupId)}/members`, { user_id: userId });
  }

  removeMember(groupId: string, userId: string): Promise<void> {
    return this.http.delete<void>(
      `/v1/groups/${encodeURIComponent(groupId)}/members/${encodeURIComponent(userId)}`,
    );
  }

  async listMembers(groupId: string): Promise<GroupMember[]> {
    const res = await this.http.get<ListEnvelope<GroupMember>>(
      `/v1/groups/${encodeURIComponent(groupId)}/members`,
    );
    return res.items ?? res.data ?? [];
  }
}

File: sdk/go/groups.go
package qeetid

import (
	"context"
	"net/http"
	"net/url"
	"strconv"
)

type Group struct {
	ID          string         `json:"id"`
	TenantID    string         `json:"tenant_id"`
	Name        string         `json:"name"`
	Description string         `json:"description,omitempty"`
	Metadata    map[string]any `json:"metadata,omitempty"`
	CreatedAt   string         `json:"created_at"`
	UpdatedAt   string         `json:"updated_at,omitempty"`
}

type CreateGroupInput struct {
	Name        string         `json:"name"`
	TenantID    string         `json:"tenant_id,omitempty"`
	Description string         `json:"description,omitempty"`
	Metadata    map[string]any `json:"metadata,omitempty"`
}

type UpdateGroupInput struct {
	Name        *string        `json:"name,omitempty"`
	Description *string        `json:"description,omitempty"`
	Metadata    map[string]any `json:"metadata,omitempty"`
}

type GroupMember struct {
	UserID  string `json:"user_id"`
	GroupID string `json:"group_id"`
	AddedAt string `json:"added_at"`
}

type GroupPage struct {
	Data       []Group
	NextCursor string
}

type Groups struct{ http *httpClient }

func (r *Groups) Create(ctx context.Context, in CreateGroupInput) (*Group, error) {
	var out Group
	err := r.http.do(ctx, http.MethodPost, "/v1/groups", nil, in, &out, false)
	return &out, err
}

func (r *Groups) Get(ctx context.Context, id string) (*Group, error) {
	var out Group
	err := r.http.do(ctx, http.MethodGet, "/v1/groups/"+url.PathEscape(id), nil, nil, &out, true)
	return &out, err
}

func (r *Groups) Update(ctx context.Context, id string, in UpdateGroupInput) (*Group, error) {
	var out Group
	err := r.http.do(ctx, http.MethodPatch, "/v1/groups/"+url.PathEscape(id), nil, in, &out, false)
	return &out, err
}

func (r *Groups) Delete(ctx context.Context, id string) error {
	return r.http.do(ctx, http.MethodDelete, "/v1/groups/"+url.PathEscape(id), nil, nil, nil, true)
}

func (r *Groups) List(ctx context.Context, params ListParams) (*GroupPage, error) {
	q := url.Values{}
	if params.Tenant != "" {
		q.Set("tenant", params.Tenant)
	}
	if params.Limit > 0 {
		q.Set("limit", strconv.Itoa(params.Limit))
	}
	if params.Cursor != "" {
		q.Set("cursor", params.Cursor)
	}
	var env struct {
		Items      []Group `json:"items"`
		Data       []Group `json:"data"`
		NextCursor string  `json:"next_cursor"`
	}
	if err := r.http.do(ctx, http.MethodGet, "/v1/groups", q, nil, &env, true); err != nil {
		return nil, err
	}
	items := env.Items
	if items == nil {
		items = env.Data
	}
	return &GroupPage{Data: items, NextCursor: env.NextCursor}, nil
}

func (r *Groups) AddMember(ctx context.Context, groupID, userID string) error {
	body := map[string]string{"user_id": userID}
	return r.http.do(ctx, http.MethodPost, "/v1/groups/"+url.PathEscape(groupID)+"/members", nil, body, nil, false)
}

func (r *Groups) RemoveMember(ctx context.Context, groupID, userID string) error {
	return r.http.do(ctx, http.MethodDelete, "/v1/groups/"+url.PathEscape(groupID)+"/members/"+url.PathEscape(userID), nil, nil, nil, true)
}

func (r *Groups) ListMembers(ctx context.Context, groupID string) ([]GroupMember, error) {
	var env struct {
		Items []GroupMember `json:"items"`
		Data  []GroupMember `json:"data"`
	}
	if err := r.http.do(ctx, http.MethodGet, "/v1/groups/"+url.PathEscape(groupID)+"/members", nil, nil, &env, true); err != nil {
		return nil, err
	}
	if env.Items != nil {
		return env.Items, nil
	}
	return env.Data, nil
}
